### PR TITLE
solution

### PR DESCRIPTION
**Pull request by Fridrif27**
[DEMO LINK](https://fridrif27.github.io/react_todo-app-add-and-delete/)

### CODE REVIEW

**Review by Welbrn — APPROVED** [review]
Good job!

File: src/App.tsx [hunk 1]
```diff
@@ -1,26 +1,259 @@
...
+import React, { useEffect, useRef, useState } from 'react';
+import { deleteTodo, getTodos, addTodo, USER_ID } from './api/todos';
+import { Todo } from './types/Todo';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { Error } from './components/Errors';
 
-const USER_ID = 0;
+export enum Filter {
```
Comment (VolodymyrKirichenko) [inline src/App.tsx:12]: move all ur enums to separate directory

File: src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,259 @@
...
+  const [sortBy, setSortBy] = useState('all');
+  const [errorMessage, setErrorMessage] = useState<string | null | boolean>(
+    null,
+  );
+  const [deleteTodoId, setDeleteTodoId] = useState<number | null>(null);
+  const [isLoadingChange, setIsLoadingChange] = useState(false);
+  const [cleanCompleted, setCleanCompleted] = useState<boolean>(false);
+  const [newTask, setNewTask] = useState<string>('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isSubmiting, setIsSubmiting] = useState<boolean>(false);
```
Comment (VolodymyrKirichenko) [inline src/App.tsx:31]: 1) Create enum for all the error `FiltersType.Default = ''`
2) Create enum for statuses
3) if u set (string, number, logical value) as the initial value - u don't need a generic type for this
```suggestion
  const [query, setQuery] = useState('');
  const [todos, setTodos] = useState<Todo[]>([]);
  const [itemsLeft, setItemsLeft] = useState(0);
  const [sortBy, setSortBy] = useState<FiltersType>(FiltersType.All);
  const [errorMessage, setErrorMessage] = useState<ErrorType>(ErrorType.Default);
  const [deleteTodoId, setDeleteTodoId] = useState<number | null>(null);
  const [isLoadingChange, setIsLoadingChange] = useState(false);
  const [cleanCompleted, setCleanCompleted] = useState(false);
  const [newTask, setNewTask] = useState('');
  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
  const [isSubmiting, setIsSubmiting] = useState(false);
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,26 +1,259 @@
...
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  function handleSubmit(e: React.FormEvent<HTMLFormElement>) {
+    e.preventDefault();
+    if (query.trim() !== '') {
```
Comment (VolodymyrKirichenko) [inline src/App.tsx:37]: the same, but more readable

```suggestion
    if (query.trim().length) {
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,259 @@
...
-  if (!USER_ID) {
-    return <UserWarning />;
+  const [query, setQuery] = useState<string>('');
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [itemsLeft, setItemsLeft] = useState<number>(0);
+  const [sortBy, setSortBy] = useState('all');
+  const [errorMessage, setErrorMessage] = useState<string | null | boolean>(
+    null,
+  );
+  const [deleteTodoId, setDeleteTodoId] = useState<number | null>(null);
```
Comment (VolodymyrKirichenko) [inline src/App.tsx:26]: do you need this state? Why?
you can get this id in function arguments

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,259 @@
...
+
+    return () => {
+      if (cleanup instanceof Function) {
+        cleanup();
+      }
+    };
+  }, [deleteTodoId, todos]);
+
+  useEffect(() => {
+    async function deleteTodoFromServer() {
```
Comment (VolodymyrKirichenko) [inline src/App.tsx:140]: `deleteTodoFromServer` of `deleteTodosFromServer`?

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,259 @@
...
+            setTodos(sortedByActive);
+            break;
+          case Filter.Completed:
+            const sortedByCompleted: Todo[] = todosFromServer.filter(
+              todo => todo.completed,
+            );
+
+            setTodos(sortedByCompleted);
+            break;
+        }
```
Comment (VolodymyrKirichenko) [inline src/App.tsx:214]: move to separate function

**Review by VolodymyrKirichenko — CHANGES_REQUESTED** [review]
Good job, but let's offload your main component a bit and improve the code 🥹

**Pull request closed without merging**